WordPress/wordpress-develop
# Global styles: preload user global styles based on user caps

## What

Syncs https://github.com/WordPress/gutenberg/pull/66541

For the Post Editor, preload `'/wp/v2/global-styles/' . $global_styles_id` with a context corresponding to user caps, that is, `'edit'` for users that can edit global styles, and `'view'` for everyone else.

## Why
Preloading the global styles endpoint according to role context means that admins and non admins, e.g., editors, avoid unnecessary clientside requests.

Different [clientside requests fire](https://github.com/WordPress/gutenberg/blob/update%2Fglobal-styles-preload-paths-remove-6.6-compat/packages/editor/src/components/global-styles-provider/index.js#L87-L87) for each.

## How?
Use `current_user_can( 'edit_theme_options' )` to test for global styles edit caps. This [matches post type caps](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/post.php#L494).

## Testing Instructions

1. Create two users: one admin, another editor or other role that can edit posts.
2. For each user, fire up this branch and head to the post editor.
3. Check that `global-styles` endpoints aren't fetched client side, but are, rather, preloaded via [rest_preload_api_request](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api.php#L2918)
4. As admin, check the same for the site editor to make sure this patch does not cause any regressions there.

For admins in the POST and SITE editors the following URLS should be preloaded:

```
 "/wp/v2/global-styles/<id>"  - "OPTIONS"
"/wp/v2/global-styles/themes/twentytwentyfive?context=view"  - "GET"
"/wp/v2/global-styles/themes/twentytwentyfive/variations?context=view"  - "GET"
"/wp/v2/global-styles/<id>5?context=edit"  - "GET"
```


For editors in the POST editor:

```
"/wp/v2/global-styles/<id>"  - "OPTIONS"
"/wp/v2/global-styles/themes/twentytwentyfive?context=view"  - "GET"
"/wp/v2/global-styles/themes/twentytwentyfive/variations?context=view"  - "GET"
"/wp/v2/global-styles/<id>?context=view"  - "GET"
```


Trac ticket: https://core.trac.wordpress.org/ticket/62322

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**
